api/v1alpha4/cluster_types.go

```diff
@@ -96,6 +96,9 @@ type ControlPlaneTopology struct {
 	Metadata ObjectMeta `json:"metadata,omitempty"`
 
 	// Replicas is the number of control plane nodes.
+	// If the value is nil, the ControlPlane object is created without the number of Replicas
+	// and it's assumed that the control plane controller does not implement support for this field.
+	// +optional
```

Comment:
Do we want to ignore this field if this is specified for a CP provider which does not support the Replicas field, and adding a comment around it?
```suggestion
     // When specified against a control plane provider that lacks support for this field, this value will be ignored.
     // +optional
```

Reply:
I'm not sure if we want to handle that case. I think it's cleaner that we expect the ControlPlane provider to have this field when it's specified in the ClusterClass.

Reply:
@srm09 how do you envision inspecting a provider for lack of support of this field?
I would like to avoid inspecting the OpenAPI spec of providers, at least in the first iteration...

Reply:
I was hoping for something of this sort to be in place, https://github.com/kubernetes-sigs/cluster-api/blob/8131b65f00606a3dd625caa1d87a330c6d7bc188/controllers/machine_controller_phases.go#L301-L309
Especially, line 304, which specifically considers failureDomains not being present as a no-op.

Reply:
@srm09 The problem is that the fact that we cannot find it in an object does not mean it's not in the CRD/ schema.

Reply:
I think I understand what you are saying. This might not be the right way to do it, and we may have to inspect the spec of the CP provider like Fabrizio mentioned.

Keeping the implementation details aside, ignoring if not present, would make the end-user UX better. I do not mind tackling this and other API spec validations together, but it is something that we should strongly consider since the primary motivation behind CC is to better the UX.

Reply:
@srm09 I wonder if this would really make the UX better. So we assume: user has set replicas even though the ControlPlane does not support it. We have two options:

1. try to find a way to not set the value > Imho this would be pretty surprising for the user that even though he has set the value and we don't throw an error the configuration has no effect
2. set the replicas on the ControlPlane (when it's set in the topology) > I assume this will throw an error when we try to create the ControlPlane as it's not part of the spec, then we can return an error in our reconcile loop > The user can see that his configuration is wrong (and probably fixes the configuration error)

I personally think the second option is the better UX as we give feedback that the configuration is wrong so that it can be fixed. In the first option we just ignore the error and there's no way for the user to get feedback.

But as Vince said below let's open an issue where we can discuss this and find a consensus :)

Reply:
#5019